### PR TITLE
Add support for JNI based decompression for zstd files

### PR DESCRIPTION
**Pull request by gauravkm**
Initial results and estimates point to 10% improvement in CPU by using JNI.

The feature is controlled by a flag, that is false by default.

I tried to split the pull requests but there is no way to stack them across different forks.

### CODE REVIEW

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkZstdJniDecompression.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+            sb.append(alphaNumericString
+                    .charAt(index));
+        }
+
+        return sb.toString();
+    }
+
+    static class Unit
+    {
+        byte[] srcBytes;
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkZstdJniDecompression.java]: Could you follow our developer guideline (https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting) and clean up this patch including:
- Spell out every word (e.g., src -> source)
- Provide meaningful names to variables (e.g., i -> length)
- We don't use ArrayList, LinkedList, and other Java native containers unless necessary (e.g., mutability or nulls). Instead, we use Guava ImmutableSet/Map/List

File: presto-orc/src/test/java/com/facebook/presto/orc/TestJniZstdDecompression.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import com.facebook.presto.orc.zstd.ZstdJniCompressor;
+import com.facebook.presto.orc.zstd.ZstdJniDecompressor;
+import com.facebook.presto.testing.assertions.Assert;
+import org.testng.annotations.Test;
+
+import java.util.Random;
+
+public class TestJniZstdDecompression
+{
+    private static final long MAX_BUFFER_SIZE = 4 * 1024L * 1024 * 1024;
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestJniZstdDecompression.java]: Let's use `DataSize` type

File: presto-orc/src/test/java/com/facebook/presto/orc/TestJniZstdDecompression.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    private final ZstdJniDecompressor decompressor = new ZstdJniDecompressor(new OrcDataSourceId("test"), MAX_BUFFER_SIZE);
+
+    @Test
+    public void testDecompression()
+            throws OrcCorruptionException
+    {
+        byte[] srcBytes = generateRandomBytes();
+        byte[] compressedBytes = new byte[1024 * 1024];
+        int size = compressor.compress(srcBytes, 0, srcBytes.length, compressedBytes, 0, compressedBytes.length);
+        int outputSize = decompressor.decompress(compressedBytes, 0, size,
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestJniZstdDecompression.java]: Also, we enforce one parameter per line

**Comment by gauravkm** [comment]
> Could you remove binary files like sample_zstd? Just randomly generate them with as temporary files and read them back

This are really ORC files. Not possible to generate them randomly. I can rename them to say sample_orc

File: presto-orc/src/test/java/com/facebook/presto/orc/TestJniZstdDecompression.java (hunk 2 above)
Comment (gauravkm) [reply presto-orc/src/test/java/com/facebook/presto/orc/TestJniZstdDecompression.java]: (y)

File: presto-orc/src/test/java/com/facebook/presto/orc/TestJniZstdDecompression.java (hunk 3 above)
Comment (mbasmanova) [reply presto-orc/src/test/java/com/facebook/presto/orc/TestJniZstdDecompression.java]: @gauravkm Gaurav, you can generate ORC files on the fly. See com.facebook.presto.orc.OrcTester#writeOrcColumnHive

**Comment by mbasmanova** [comment]
@gauravkm 

> This are really ORC files. Not possible to generate them randomly. I can rename them to say sample_orc

Gaurav, you can generate ORC files on the fly. See com.facebook.presto.orc.OrcTester#writeOrcColumnHive

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java [hunk 4]
```diff
@@ -573,12 +573,14 @@ public synchronized void checkConnectorWrite(ConnectorId connectorId)
         private synchronized ListenableFuture<?> abortInternal()
         {
             // the callbacks in statement performed on another thread so are safe
-            return nonCancellationPropagating(Futures.allAsList(Stream.concat(
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java]: Unrelated change?

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcDecompressor.java [hunk 5]
```diff
@@ -42,7 +49,12 @@
             case LZ4:
                 return Optional.of(new OrcLz4Decompressor(orcDataSourceId, bufferSize));
             case ZSTD:
-                return Optional.of(new OrcZstdDecompressor(orcDataSourceId, bufferSize));
+                if (zstdJniDecompressionEnabled) {
+                    return Optional.of(new ZstdJniDecompressor(orcDataSourceId, bufferSize));
+                }
+                else {
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcDecompressor.java]: `else {` is redundant

File: presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public int decompress(byte[] srcBytes, int offset, int length, OutputBuffer output)
+            throws OrcCorruptionException
+    {
+        try {
+            long uncompressedLength = ZstdDecompressor.getDecompressedSize(srcBytes, offset, length);
+            if (uncompressedLength > this.maxBufferSize) {
+                throw new OrcCorruptionException(orcDataSourceId, "Zstd requires buffer (%s) larger than max size (%s)", uncompressedLength, maxBufferSize);
+            }
+            byte[] dstBuffer = output.initialize(toIntExact(uncompressedLength));
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java]: spell out dst

File: presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        this.maxBufferSize = maxBufferSize;
+    }
+
+    @Override
+    public int decompress(byte[] srcBytes, int offset, int length, OutputBuffer output)
+            throws OrcCorruptionException
+    {
+        try {
+            long uncompressedLength = ZstdDecompressor.getDecompressedSize(srcBytes, offset, length);
+            if (uncompressedLength > this.maxBufferSize) {
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java]: remove `this.`

File: presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                throw new OrcCorruptionException(orcDataSourceId, "Zstd requires buffer (%s) larger than max size (%s)", uncompressedLength, maxBufferSize);
+            }
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java]: break a new line after }

File: presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    private final long maxBufferSize;
+
+    public ZstdJniDecompressor(OrcDataSourceId orcDataSourceId, long maxBufferSize)
+    {
+        this.orcDataSourceId = requireNonNull(orcDataSourceId, "orcDataSourceId is null");
+        this.maxBufferSize = maxBufferSize;
+    }
+
+    @Override
+    public int decompress(byte[] srcBytes, int offset, int length, OutputBuffer output)
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java]: spell out src

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReadersWithZstd.java [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    private static final int ROWS = 10_000_000;
+    private static final int MAX_STRING = 10;
+    private static final List<?> NULL_VALUES = Collections.nCopies(ROWS, null);
+
+    @Benchmark
+    public Object readBlocksWithoutJni(BenchmarkData data)
+            throws Throwable
+    {
+        OrcBatchRecordReader recordReader = data.createRecordReader(false);
+        List<Block> blocks = new ArrayList<>();
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReadersWithZstd.java]: Use ImmutableList.builder

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReadersWithZstd.java [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+        }
+        return blocks;
+    }
+
+    @Benchmark
+    public Object readBlocksWithJni(BenchmarkData data)
+            throws Throwable
+    {
+        OrcBatchRecordReader recordReader = data.createRecordReader(true);
+        List<Block> blocks = new ArrayList<>();
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReadersWithZstd.java]: same

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkZstdJniDecompression.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+            });
+            assertEquals(outputSize, unit.sourceLength);
+        }
+    }
+
+    private static List<Unit> generateWorkload()
+    {
+        ImmutableList.Builder<Unit> builder = new ImmutableList.Builder<>();
+        for (int i = 0; i < 10; i++) {
+            byte[] srcBytes = getAlphaNumericString(sourceLength).getBytes();
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkZstdJniDecompression.java]: spell out src

File: presto-orc/src/test/java/com/facebook/presto/orc/TestZstdJniDecompression.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    private final ZstdJniDecompressor decompressor = new ZstdJniDecompressor(new OrcDataSourceId("test"), MAX_BUFFER_SIZE.toBytes());
+
+    @Test
+    public void testDecompression()
+            throws OrcCorruptionException
+    {
+        byte[] sourceBytes = generateRandomBytes();
+        byte[] compressedBytes = new byte[1024 * 1024];
+        int size = compressor.compress(sourceBytes, 0, sourceBytes.length, compressedBytes, 0, compressedBytes.length);
+        int outputSize = decompressor.decompress(compressedBytes, 0, size,
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestZstdJniDecompression.java]: one param per line

File: presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+            throws OrcCorruptionException
+    {
+        try {
+            long uncompressedLength = ZstdDecompressor.getDecompressedSize(srcBytes, offset, length);
+            if (uncompressedLength > this.maxBufferSize) {
+                throw new OrcCorruptionException(orcDataSourceId, "Zstd requires buffer (%s) larger than max size (%s)", uncompressedLength, maxBufferSize);
+            }
+            byte[] dstBuffer = output.initialize(toIntExact(uncompressedLength));
+            long size = Zstd.decompressByteArray(dstBuffer, 0, dstBuffer.length, srcBytes, offset, length);
+            return (int) size;
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java]: Do we wanna test on `Zstd.isError(size)`?

File: presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        this.orcDataSourceId = requireNonNull(orcDataSourceId, "orcDataSourceId is null");
+        this.maxBufferSize = maxBufferSize;
+    }
+
+    @Override
+    public int decompress(byte[] srcBytes, int offset, int length, OutputBuffer output)
+            throws OrcCorruptionException
+    {
+        try {
+            long uncompressedLength = ZstdDecompressor.getDecompressedSize(srcBytes, offset, length);
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java]: This reuses aircompressor. So two options maybe:
1. Share the same codepath with `OrcZstdDecompressor` and pass in the jni flag to decide which decompressor (jni or airlift) to use within `OrcZstdDecompressor`, or
2. Keep this class and remove aircompressor dependency.

File: presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java (hunk 4 above)
Comment (gauravkm) [reply presto-main/src/main/java/com/facebook/presto/transaction/InMemoryTransactionManager.java]: Will put a separate diff if required. For some reason, this would not allow Idea to compile the code.

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReadersWithZstd.java [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+        }
+
+        @TearDown
+        public void tearDown()
+                throws IOException
+        {
+            deleteRecursively(temporaryDirectory.toPath(), ALLOW_INSECURE);
+        }
+
+        protected final List<?> createValues()
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReadersWithZstd.java]: nit: we don't usually use `final` on method. (Also, as a side note: we don't use `final` for tmp variables)

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReadersWithZstd.java [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+                case ALL:
+                    return NULL_VALUES;
+                case PARTIAL:
+                    return IntStream.range(0, ROWS).mapToObj(i -> i % 2 == 0 ? createValue() : null).collect(toList());
+                default:
+                    return IntStream.range(0, ROWS).mapToObj(i -> createValue()).collect(toList());
+            }
+        }
+
+        private final Object createValue()
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReadersWithZstd.java]: same, remove final

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkZstdJniDecompression.java [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+        }
+        return builder.build();
+    }
+
+    private OrcDecompressor createOrcDecompressor(boolean zstdJniDecompressionEnabled)
+    {
+        return new OrcZstdDecompressor(new OrcDataSourceId("orc"), sourceLength, zstdJniDecompressionEnabled);
+    }
+
+    static String getAlphaNumericString(int length)
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkZstdJniDecompression.java]: private

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkZstdJniDecompression.java [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+            int arrayIndex = (int) (alphaNumericString.length() * Math.random());
+
+            stringBuilder.append(alphaNumericString.charAt(arrayIndex));
+        }
+        return stringBuilder.toString();
+    }
+
+    static class Unit
+    {
+        byte[] sourceBytes;
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkZstdJniDecompression.java]: put `final` to all of them

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcReaderTestingUtils.java [hunk 20]
```diff
@@ -29,4 +29,13 @@ public static OrcReaderOptions createDefaultTestConfig()
                 new DataSize(1, MEGABYTE),
                 false);
     }
+
+    public static OrcReaderOptions createOrcReaderOptionsAndSetZstdJni(boolean zstdJniDecompressionEnabled)
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcReaderTestingUtils.java]: Maybe just call it `createTestingReaderOptions`

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcReaderTestingUtils.java [hunk 21]
```diff
@@ -29,4 +29,13 @@ public static OrcReaderOptions createDefaultTestConfig()
                 new DataSize(1, MEGABYTE),
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcReaderTestingUtils.java]: Also call this function `createTestingReaderOptions` and it should delegate to the one below with `zstdJniDecompressionEnabled = false`.

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java [hunk 22]
```diff
@@ -1701,6 +1701,12 @@ public static DataSize writeOrcColumnHive(File outputFile, Format format, Compre
         return writeOrcColumnsHive(outputFile, format, compression, ImmutableList.of(type), ImmutableList.of(values));
     }
 
+    public static void writeOrcColumnPresto(File outputFile, Format format, CompressionKind compression, Type type, List<?> values)
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java]: move this closer to the other `writeOrcColumnPresto`

File: presto-orc/src/test/java/com/facebook/presto/orc/TestZstdJniDecompression.java [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+                        return new byte[size];
+                    }
+
+                    @Override
+                    public byte[] grow(int size)
+                    {
+                        throw new RuntimeException();
+                    }
+                });
+        Assert.assertEquals(outputSize, sourceBytes.length);
```
Comment (highker) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestZstdJniDecompression.java]: Shall we assert on the data equal. Size equality seems a bit weak.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcZstdDecompressor.java [hunk 24]
```diff
@@ -24,12 +25,19 @@
...
-    private final ZstdDecompressor decompressor = new ZstdDecompressor();
+    private final Decompressor decompressor;
 
-    public OrcZstdDecompressor(OrcDataSourceId orcDataSourceId, int maxBufferSize)
+    public OrcZstdDecompressor(OrcDataSourceId orcDataSourceId, int maxBufferSize, boolean zstdJniDecompressionEnabled)
     {
         this.orcDataSourceId = requireNonNull(orcDataSourceId, "orcDataSourceId is null");
         this.maxBufferSize = maxBufferSize;
+        if (zstdJniDecompressionEnabled) {
+            this.decompressor = (input, inputOffset, inputLength, output, outputOffset, maxOutputLength) -> (int) Zstd.decompressByteArray(output, 0, maxOutputLength, input, inputOffset, inputLength);
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcZstdDecompressor.java]: nit
```java
this.decompressor = (input, inputOffset, inputLength, output, outputOffset, maxOutputLength) -> {
    long size = Zstd.decompressByteArray(output, 0, maxOutputLength, input, inputOffset, inputLength);
    if (Zstd.isError(size)) {
        throw new RuntimeException(Zstd.getErrorName(size));
    }
    return toIntExact(size);
};

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcDecompressor.java [hunk 25]
```diff
@@ -28,6 +28,12 @@
 
     static Optional<OrcDecompressor> createOrcDecompressor(OrcDataSourceId orcDataSourceId, CompressionKind compression, int bufferSize)
             throws OrcCorruptionException
+    {
+        return createOrcDecompressor(orcDataSourceId, compression, bufferSize, false);
+    }
+
+    static Optional<OrcDecompressor> createOrcDecompressor(OrcDataSourceId orcDataSourceId, CompressionKind compression, int bufferSize, boolean zstdJniDecompressionEnabled)
+            throws OrcCorruptionException
     {
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcDecompressor.java]: check state here `zstdJniDecompressionEnabled` is true only if `compression` is `ZSTD`

Comment (highker) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcDecompressor.java]: Actually no; no need to check state

File: presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java (hunk 14 above)
Comment (gauravkm) [reply presto-orc/src/main/java/com/facebook/presto/orc/zstd/ZstdJniDecompressor.java]: Yes, we should I missed this part. I am working on making this change. Thanks for looking at it again. Will make the other changes as well.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcZstdDecompressor.java (hunk 24 above)
Comment (gauravkm) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcZstdDecompressor.java]: I am not quite sure if we should throw RuntimeException or MalformedInputException. 

Currently decompress catches the MalformInputException (which is a runtime exception) and rethrows it as OrcCorruptException which is a checked exception.

If we throw Runtime exception, then this behavior changes.
